druid-shell/src/platform/mac/text_input.rs

```diff
@@ -0,0 +1,679 @@
+            let second_grapheme = edit_lock.slice(middle_idx..selection.downstream_index());
+            let new_string = format!("{}{}", second_grapheme, first_grapheme);
+            // replace_range should automatically set selection to end of inserted range
+            edit_lock.replace_range(selection.to_range(), &new_string);
+        }
+        "capitalizeWord:" => {
+            // this command expands the selection to words, and then applies titlecase to that selection
+            // not actually sure what keyboard shortcut corresponds to this or the other case changing commands,
+            // but textedit seems to have this behavior when this action is simulated by editing DefaultKeyBinding.dict.
+            edit_lock.handle_action(TextInputAction::SelectWord);
+            edit_lock.handle_action(TextInputAction::TitlecaseSelection)
+        }
+        "lowercaseWord:" => {
+            // this command expands the selection to words, and then applies uppercase to that selection
+            edit_lock.handle_action(TextInputAction::SelectWord);
+            edit_lock.handle_action(TextInputAction::LowercaseSelection);
+        }
+        "uppercaseWord:" => {
+            // this command expands the selection to words, and then applies uppercase to that selection
+            edit_lock.handle_action(TextInputAction::SelectWord);
+            edit_lock.handle_action(TextInputAction::UppercaseSelection);
+        }
+        "insertDoubleQuoteIgnoringSubstitution:" => {
+            edit_lock.handle_action(TextInputAction::InsertDoubleQuoteIgnoringSmartQuotes)
+        }
+        "insertSingleQuoteIgnoringSubstitution:" => {
+            edit_lock.handle_action(TextInputAction::InsertSingleQuoteIgnoringSmartQuotes)
+        }
+        "cancelOperation:" => edit_lock.handle_action(TextInputAction::Cancel),
+        // "deleteToMark:" => {}          // TODO(lord): selectToMark, then delete selection. also puts selection in yank buffer
+        // "selectToMark:" => {}          // TODO(lord): extends the selection to include the mark
+        // "setMark:" => {}               // TODO(lord): remembers index in document (but what about grapheme clusters??)
+        // "swapWithMark:" => {}          // TODO(lord): swaps current and stored mark indices
+        // "yank:" => {}                  // TODO(lord): triggered with control-y, inserts in the text previously killed deleteTo*OfParagraph
+        "transposeWords:" => {} // textedit doesn't support, so neither will we
+        "changeCaseOfLetter:" => {} // textedit doesn't support, so neither will we
+        "indent:" => {}         // textedit doesn't support, so neither will we
+        "insertContainerBreak:" => {} // textedit and pages don't seem to support, so neither will we
+        "quickLookPreviewItems:" => {} // don't seem to apply to text editing?? hopefully
+        "complete:" => {}             // intercepted by the IME; if it gets to us, noop
```

Comment:
this one makes me do a weird face.

how and where does this get sent? do we actually receive this on return in some circumstances, or does the IME translate it to an 'insert newline' or something if there's no active IME session? If we're at all unsure it might at least be worth logging since it will be potentially a confusing thing to have to debug.

Reply:
I've, uh, never seen them get sent. I only know this behavior by changing my macOS text editing keybindings to force this to get sent to the text input system. A log is a great idea, I've added that.